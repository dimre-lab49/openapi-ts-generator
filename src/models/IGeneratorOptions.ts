export interface IGeneratorOptions {
  generateSubTypeFactory?: boolean;
  modelFolder: string;
  enumTSFile: string;
  templateFolder: string;
  generateBaseClass?: boolean;
  generateBarrelFiles?: boolean;
  generateClasses?: boolean;
  generateValidatorFile?: boolean;
  baseModelFileName?: string;
  subTypeFactoryFileName?: string;
  validatorsFileName?: string;
  exclude?: Array<string | RegExp>;
  enumI18NHtmlFile?: string;
  enumLanguageFiles?: string[];
  modelModuleName?: string;
  enumModuleName?: string;
  enumRef?: string;
  subTypePropertyName?: string;
  namespacePrefixesToRemove?: string[];
  typeNameSuffixesToRemove?: string[];
  typesToFilter?: string[];
  propertiesToFilter?: string[];
  sortModelProperties?: boolean;
  sortEnumTypes?: boolean;

  templates?: {
    validators?: string;
    baseModel?: string;
    models?: string;
    subTypeFactory?: string;
    formGroupFacTemplate?: string;
    barrel?: string;
    enum?: string;
    enumLanguage?: string;
    simpleEnum?: string;
  };
}
