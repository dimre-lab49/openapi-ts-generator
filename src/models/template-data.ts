export interface ITemplateData {
  entities?: IEntity[];
  paths: string[];
}
export interface IEntity {
  name: string;
  importTypes: IImportType[];
  valueProperties: IValueProperty[];
  referenceProperties: IReferenceProperty[];
}
export interface IImportType {
  kebabCasedTypeName: string;
  name: string;
}
export interface IValueProperty {
  name: string;
  snakeCaseName: string;
  typeScriptType?: string;
  required: boolean;
  maxLength?: number;
  minLength?: Int32Array;
}

export interface IReferenceProperty {
  name: string;
  snakeCaseName: string;
  referenceTypeName: string;
  isArray: boolean;
  required: boolean;
}
