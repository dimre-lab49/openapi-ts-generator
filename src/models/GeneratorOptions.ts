import { resolve } from 'path';
import { IGeneratorOptions } from './IGeneratorOptions';

export class GeneratorOptions {
  public generateBaseClass: boolean;
  public templateFolder: string;
  public modelFolder: string;
  public enumTSFile: string;
  public generateBarrelFiles: boolean;
  public generateClasses: boolean;
  public generateValidatorFile: boolean;
  public baseModelFileName: string;
  public subTypeFactoryFileName: string;
  public validatorsFileName: string;
  public exclude: Array<string | RegExp>;
  public enumI18NHtmlFile: string | null;
  public enumLanguageFiles: string[] | null;
  public modelModuleName: string | null;
  public enumModuleName: string | null;
  public enumRef: string | null;
  public subTypePropertyName: string;
  public namespacePrefixesToRemove: string[];
  public typeNameSuffixesToRemove: string[];
  public typesToFilter: string[];

  public templates: {
    validators: string;
    baseModel: string;
    models: string;
    subTypeFactory: string;
    barrel: string;
    enum: string;
    enumLanguage: string;
  };
  public generateSubTypeFactory: boolean;

  public constructor(options: IGeneratorOptions) {
    this.generateBaseClass = options.generateBaseClass || false;
    this.templateFolder = options.templateFolder;
    this.modelFolder = options.modelFolder;
    this.enumTSFile = options.enumTSFile;
    this.exclude = options.exclude || [];
    this.generateSubTypeFactory = options.generateSubTypeFactory || false;
    this.enumI18NHtmlFile = options.enumI18NHtmlFile || null;
    this.enumLanguageFiles = options.enumLanguageFiles || null;
    this.modelModuleName = options.modelModuleName || null;
    this.enumModuleName = options.enumModuleName || null;

    this.enumRef = options.enumRef || null;

    this.namespacePrefixesToRemove = options.namespacePrefixesToRemove || [];

    this.typeNameSuffixesToRemove = options.typeNameSuffixesToRemove || [];
    this.typesToFilter = options.typesToFilter || [];

    this.baseModelFileName = options.baseModelFileName || 'base-model.ts';
    this.generateBarrelFiles = options.generateBarrelFiles || true;
    this.generateClasses = options.generateClasses || true;
    this.generateValidatorFile = options.generateValidatorFile || false;
    this.subTypeFactoryFileName = options.subTypeFactoryFileName || 'sub-type-factory.ts';
    this.subTypePropertyName = options.subTypePropertyName || '$type';
    this.validatorsFileName = options.validatorsFileName || 'validators.ts';

    const template = options.templates || {};
    this.templates = {
      barrel: template.barrel || `${this.templateFolder}/generate-barrel-ts.hbs`,
      baseModel: template.baseModel || `${this.templateFolder}/generate-base-model-ts.hbs`,
      enum: template.enum || `${this.templateFolder}/generate-enum-ts.hbs`,
      enumLanguage: template.enumLanguage || `${this.templateFolder}/generate-enum-i18n-html.hbs`,
      models: template.models || `${this.templateFolder}/generate-model-ts.hbs`,
      subTypeFactory: template.subTypeFactory || `${this.templateFolder}/generate-sub-type-factory-ts.hbs`,
      validators: template.validators || `${this.templateFolder}/generate-validators-ts.hbs`,
    };
  }
}
