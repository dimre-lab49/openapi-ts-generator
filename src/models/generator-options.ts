import { IEntity, IReferenceProperty, IValueProperty } from './template-data';

export interface IGeneratorOptions {
  outputPath: string;
  openApiJsonUrl: string;
  messageFormat?: string | null;
  angular?: boolean | null;
  typeFilterCallBack?: (entity: IEntity, index: number, array: IEntity[]) => boolean;
  valuePropertyTypeFilterCallBack?: (valueProperty: IValueProperty, index: number, array: IValueProperty[]) => boolean;
  referencePropertyTypeFilterCallBack?: (
    referenceProperty: IReferenceProperty,
    index: number,
    array: IReferenceProperty[],
  ) => boolean;
  templates?: ITemplates | null;
}
export interface ITemplates {
  model: string;
  formGroupFactory: string;
  modelProperties: string;
  barrel: string;
  enum: string;
}
export function defaultFilter(
  value: IEntity | IValueProperty | IReferenceProperty,
  index: number,
  array: IEntity[] | IValueProperty[] | IReferenceProperty[],
) {
  return true;
}

export function setGeneratorOptionDefaults(options: IGeneratorOptions): IGeneratorOptions {
  options.angular = options.angular ?? true;
  options.messageFormat = options.messageFormat ?? 'json';
  options.typeFilterCallBack = options.typeFilterCallBack ?? defaultFilter;
  options.valuePropertyTypeFilterCallBack = options.valuePropertyTypeFilterCallBack ?? defaultFilter;
  options.referencePropertyTypeFilterCallBack = options.referencePropertyTypeFilterCallBack ?? defaultFilter;
  options.templates = {
    ...options.templates,
    model: options.templates?.model ?? './templates/model.ts.hbs',
    formGroupFactory: options.templates?.formGroupFactory ?? './templates/form-group-factory.ts.hbs',
    barrel: options.templates?.barrel ?? './templates/index.ts.hbs',
    enum: options.templates?.barrel ?? './templates/enum.ts.hbs',
    modelProperties: options.templates?.barrel ?? './templates/model-properties.ts.hbs',
  };
  return options;
}
