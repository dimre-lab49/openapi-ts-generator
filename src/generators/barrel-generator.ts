import { readdirSync } from 'fs';
import _ = require('lodash');
import { IGeneratorOptions } from '../models/generator-options';
import { ITemplateData } from '../models/template-data';
import { BaseGenerator } from './base-generator';

export class BarrelGenerator extends BaseGenerator<{ fileNames: string[] }> {
  private readonly tsRegex = /.ts$/;
  constructor(options: IGeneratorOptions) {
    super(options, options.templates?.barrel);
  }

  public generate(templateData: ITemplateData): string | null {
    const fileNames = readdirSync(this.generatorOptions.outputPath).map(value => value.replace(this.tsRegex, ''));
    return super.generateFile(`${this.generatorOptions.outputPath}/index.ts`, { fileNames });
  }
}
