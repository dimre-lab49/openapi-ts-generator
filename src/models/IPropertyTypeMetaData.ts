import { ITypeMetaData } from './ITypeMetaData';
import { IValidators } from './validators';

export interface IPropertyTypeMetaData {
  name: string;
  staticFieldName: string;
  type?: ITypeMetaData;
  interfaceTypeName: string;
  typeName: string;
  namespace?: string;
  fullNamespace?: string;
  description?: string;
  hasValidation: boolean;
  isComplexType: boolean;
  isImportType: boolean;
  isUniqueImportType: boolean;
  importType?: string;
  importFile?: string;
  isEnum: boolean;
  isUniqueImportEnumType: boolean;
  importEnumType?: string;
  isArray: boolean;
  isArrayComplexType: boolean;
  arrayTypeName?: string;
  validators?: IValidators;
  enum?: string[];
}
