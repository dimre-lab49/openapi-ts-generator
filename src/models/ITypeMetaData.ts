import { IPropertyTypeMetaData } from './IPropertyTypeMetaData';

export interface ITypeMetaData {
  fileName: string;
  typeName: string;
  interfaceTypeName: string;
  namespace: string;
  fullNamespace: string;
  fullTypeName: string;
  importFile: string;
  isSubType: boolean;
  hasSubTypeProperty: boolean;
  isBaseType: boolean;
  baseType?: ITypeMetaData;
  baseImportFile?: string;
  path: string;
  pathToRoot: string;
  properties: IPropertyTypeMetaData[];
}
