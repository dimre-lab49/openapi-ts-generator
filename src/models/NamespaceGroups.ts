import { ITypeMetaData } from './ITypeMetaData';

export interface INamespaceGroups {
  [namespace: string]: ITypeMetaData[];
}
