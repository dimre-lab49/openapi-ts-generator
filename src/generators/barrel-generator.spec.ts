import { existsSync, mkdirSync } from 'fs';
import { IGeneratorOptions, setGeneratorOptionDefaults } from '../models/generator-options';
import { ITemplateData } from '../models/template-data';
import { BarrelGenerator } from './barrel-generator';

const outputPath = './jest_output';

describe('BarrelGenerator', () => {
  beforeEach(() => {
    if (!existsSync(outputPath)) {
      mkdirSync(outputPath);
    }
  });
  it('should generate', () => {
    const generator = new BarrelGenerator(
      setGeneratorOptionDefaults({
        templates: { barrel: './src/templates/index.ts.hbs' },
        outputPath,
      } as IGeneratorOptions),
    );
    const result = generator.generate({} as ITemplateData);
    expect(result).toMatchSnapshot();
  });
});
