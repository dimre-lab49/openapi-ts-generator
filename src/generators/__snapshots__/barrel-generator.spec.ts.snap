// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BarrelGenerator should generate 1`] = `
"/**
 * This file is generated by the openapi-ts-generator
 * #generate-barrel-ts.hbs
 * For issues or feature request, visit the repo: https://github.com/ikemtz/openapi-ts-generator
 * Do not edit.
 */
/* tslint:disable */
"
`;

exports[`BarrelGenerator should handle exceptions 1`] = `
Array [
  "Error executing template: C:\\\\Repos\\\\openapi-ts-generator\\\\src\\\\templates/index.ts.hbs.",
  "This is likely an issue with the template.",
  "Data: {\\"fileNames\\":[]}",
  "Goto: https://github.com/ikemtz/openapi-ts-generator to report an issue if necessary.",
  [Error: This error is to validate unit tests.],
]
`;
